### PR TITLE
Switching to pure code oidc flow, adding validation of state, showing idToken claims

### PR DESCRIPTION
**Pull request by SomkaPe**

### CODE REVIEW

File: src/main/java/com/microsoft/aad/adal4jsample/BasicFilter.java [hunk 1]
```diff
@@ -23,14 +23,8 @@
...
 import java.net.URLEncoder;
-import java.util.Date;
-import java.util.HashMap;
-import java.util.Map;
-import java.util.UUID;
-import java.util.concurrent.ExecutionException;
-import java.util.concurrent.ExecutorService;
-import java.util.concurrent.Executors;
-import java.util.concurrent.Future;
+import java.util.*;
```
Comment (polita) [inline src/main/java/com/microsoft/aad/adal4jsample/BasicFilter.java:27]: Is importing something.\* considered acceptable practice in Java?

File: src/main/java/com/microsoft/aad/adal4jsample/BasicFilter.java [hunk 2]
```diff
@@ -53,6 +49,10 @@
 
 public class BasicFilter implements Filter {
 
+    public static final String STATES = "states";
+    public static final String STATE = "state";
+    public static final Integer STATE_TTL = 180;
+    public static final String FAILED_TO_VAL_MSG = "Failed to validate data received from Authorization service - ";
```
Comment (polita) [inline src/main/java/com/microsoft/aad/adal4jsample/BasicFilter.java]: Please use full words. I recommend "FAILED_TO_VALIDATE_MESSAGE".

File: src/main/java/com/microsoft/aad/adal4jsample/BasicFilter.java [hunk 3]
```diff
@@ -53,6 +49,10 @@
 
 public class BasicFilter implements Filter {
 
+    public static final String STATES = "states";
+    public static final String STATE = "state";
+    public static final Integer STATE_TTL = 180;
```
Comment (polita) [inline src/main/java/com/microsoft/aad/adal4jsample/BasicFilter.java]: This means we're expiring things out of state after 3 hours? Why not use the same default that passport uses (1 hour)?

File: src/main/java/com/microsoft/aad/adal4jsample/BasicFilter.java [hunk 4]
```diff
@@ -64,110 +64,137 @@ public void destroy() {
...
-                                || "https".equals(request.getScheme())
-                                && request.getServerPort() == 443 ? "" : ":"
-                                + request.getServerPort())
-                        + httpRequest.getRequestURI();
+                String currentUri = httpRequest.getRequestURL().toString();
                 String fullUrl = currentUri
-                        + (httpRequest.getQueryString() != null ? "?"
-                                + httpRequest.getQueryString() : "");
-                // check if user has a session
+                        + (httpRequest.getQueryString() != null ? "?" + httpRequest.getQueryString() : "");
```
Comment (polita) [inline src/main/java/com/microsoft/aad/adal4jsample/BasicFilter.java]: Let's cache the result of httpRequest.getQueryString() so we don't have to call it twice.

File: src/main/java/com/microsoft/aad/adal4jsample/BasicFilter.java [hunk 5]
```diff
@@ -64,110 +64,137 @@ public void destroy() {
...
-                httpResponse.sendRedirect(((HttpServletRequest) request)
-                        .getContextPath() + "/error.jsp");
+                request.getRequestDispatcher("/error.jsp").forward(request, response);
             }
         }
         chain.doFilter(request, response);
     }
 
-    private AuthenticationResult getAccessTokenFromClientCredentials()
+    private void checkAccessTokenExpirationDate(HttpServletRequest httpRequest, String currentUri) throws Throwable {
```
Comment (polita) [inline src/main/java/com/microsoft/aad/adal4jsample/BasicFilter.java]: It seems like this method is doing more than just checking the access token expiration date; it's also getting the access token from the refresh token. This method should probably stick to just checking the expiration date, then the filter should call getAccessTokenFromRefreshToken if the access token is expired.

File: src/main/java/com/microsoft/aad/adal4jsample/BasicFilter.java [hunk 6]
```diff
@@ -64,110 +64,137 @@ public void destroy() {
...
+     * @param session
+     * @param stateInAuthData
+     * @throws Exception
+     */
+    private void validateState(HttpSession session, String stateInAuthData) throws Exception {
+        if (stateInAuthData == null || !isStateInSession(session, stateInAuthData)){
+            throw new Exception(FAILED_TO_VAL_MSG + "could not validate state");
+        }
+    }
+
```
Comment (polita) [inline src/main/java/com/microsoft/aad/adal4jsample/BasicFilter.java:188]: Nit: remove extra return here.

File: src/main/webapp/index.jsp [hunk 7]
```diff
@@ -1,8 +1,12 @@
...
 <body>
-	<h2>Hello World!</h2>
-	<ul>
-	<li><a href="secure/aad">Secure Page</a></li>
-	</ul>
+	<h3>Home Page</h3>
+
+	<form action="secure/aad">
+
+		<input type="submit" value="Show users in the Tenant">
```
Comment (polita) [inline src/main/webapp/index.jsp]: "tenant" with a lower-case t

File: src/main/webapp/secure/aad.jsp [hunk 8]
```diff
@@ -6,19 +6,36 @@
 </head>
 <body>
 
-	<h1>Directory - Users List</h1>
+	<h3>List of users registered in the tenant - ${tenant}</h3>
 	<p>${users}</p>
+	<br>
+	<h3>User who authorised access to graph API</h3>
```
Comment (polita) [inline src/main/webapp/secure/aad.jsp]: "Users who are authorized to access Graph API"

Comment (polita) [reply src/main/webapp/secure/aad.jsp]: Or maybe I misunderstand. Is this the contents of the id_token claims? If so, this should say "Current user" or something similar.

File: src/main/java/com/microsoft/aad/adal4jsample/BasicFilter.java (hunk 1 above)
Comment (SomkaPe) [reply src/main/java/com/microsoft/aad/adal4jsample/BasicFilter.java:27]: It doesn't have any impact on runtime performance. There are different opinions , I personally trust IDE to manage it, It takes to account number of imports, package type ( java.util is a core java package)

File: src/main/java/com/microsoft/aad/adal4jsample/BasicFilter.java (hunk 3 above)
Comment (SomkaPe) [reply src/main/java/com/microsoft/aad/adal4jsample/BasicFilter.java]: switched to 1 hour

File: src/main/java/com/microsoft/aad/adal4jsample/BasicFilter.java (hunk 2 above)
Comment (SomkaPe) [reply src/main/java/com/microsoft/aad/adal4jsample/BasicFilter.java]: ok

File: src/main/java/com/microsoft/aad/adal4jsample/BasicFilter.java (hunk 4 above)
Comment (SomkaPe) [reply src/main/java/com/microsoft/aad/adal4jsample/BasicFilter.java]: ok

File: src/main/java/com/microsoft/aad/adal4jsample/BasicFilter.java (hunk 5 above)
Comment (SomkaPe) [reply src/main/java/com/microsoft/aad/adal4jsample/BasicFilter.java]: ok

File: src/main/webapp/secure/aad.jsp (hunk 8 above)
Comment (SomkaPe) [reply src/main/webapp/secure/aad.jsp]: yes it is  id_token claims

Comment (SomkaPe) [reply src/main/webapp/secure/aad.jsp]: ok

File: src/main/webapp/index.jsp (hunk 7 above)
Comment (SomkaPe) [reply src/main/webapp/index.jsp]: ok

File: src/main/java/com/microsoft/aad/adal4jsample/BasicFilter.java (hunk 6 above)
Comment (SomkaPe) [reply src/main/java/com/microsoft/aad/adal4jsample/BasicFilter.java:188]: ok

File: src/main/java/com/microsoft/aad/adal4jsample/BasicFilter.java (hunk 3 above)
Comment (polita) [reply src/main/java/com/microsoft/aad/adal4jsample/BasicFilter.java]: thanks

**Comment by polita** [comment]
After @lovemaths takes a look, :shipit:.

**Comment by lovemaths** [comment]
:shipit:

**Pull request merged**